modules/mono/glue/GodotSharp/GodotSharp/Core/Projection.cs

```diff
@@ -0,0 +1,822 @@
+                   x.z * y.y * z.x * w.w + x.y * y.z * z.x * w.w +
+                   x.z * y.x * z.y * w.w - x.x * y.z * z.y * w.w -
+                   x.y * y.x * z.z * w.w + x.x * y.y * z.z * w.w;
+        }
+
+        public Projection FlippedY()
+        {
+            Projection proj = this;
+            proj.y = -proj.y;
+            return proj;
+        }
+
+        public real_t GetAspect()
+        {
+            Vector2 vpHe = GetViewportHalfExtends();
+            return vpHe.x / vpHe.y;
+        }
+
+        public Vector2 GetFarPlaneHalfExtends()
+        {
+            var res = GetProjectionPlane(Planes.Far).Intersect3(GetProjectionPlane(Planes.Right), GetProjectionPlane(Planes.Top));
+            return new Vector2(res.Value.x, res.Value.y);
+        }
+
+        public real_t GetFov()
+        {
+            Plane rightPlane = new Plane(x.w - x.x, y.w - y.x, z.w - z.x, -w.w + w.x).Normalized();
+            if (z.x == 0 && z.y == 0)
+            {
+                return Mathf.Rad2Deg(Mathf.Acos(Mathf.Abs(rightPlane.Normal.x))) * (real_t)2.0;
+            }
+            else
+            {
+                Plane leftPlane = new Plane(x.w + x.x, y.w + y.x, z.w + z.x, w.w + w.x).Normalized();
+                return Mathf.Rad2Deg(Mathf.Acos(Mathf.Abs(leftPlane.Normal.x))) + Mathf.Rad2Deg(Mathf.Acos(Mathf.Abs(rightPlane.Normal.x)));
+            }
+        }
+
+        public static real_t GetFovy(real_t fovx, real_t aspect)
+        {
+            return Mathf.Rad2Deg(Mathf.Atan(aspect * Mathf.Tan(Mathf.Deg2Rad(fovx) * (real_t)0.5)) * (real_t)2.0);
+        }
+
+        public real_t GetLodMultiplier()
+        {
+            if (IsOrthorgonal())
+            {
+                return GetViewportHalfExtends().x;
+            }
+            else
+            {
+                real_t zn = GetZNear();
+                real_t width = GetViewportHalfExtends().x * (real_t)2.0;
+                return (real_t)1.0 / (zn / width);
+            }
+        }
+
+        public int GetPixelsPerMeter(int forPixelWidth)
+        {
+            Vector3 result = Xform(new Vector3(1, 0, -1));
+
+            return (int)((result.x * (real_t)0.5 + (real_t)0.5) * forPixelWidth);
+        }
+
+        public Plane GetProjectionPlane(Planes plane)
+        {
+            Plane newPlane = plane switch
+            {
+                Planes.Near => new Plane(x.w + x.z, y.w + y.z, z.z + z.y, w.w + w.z),
+                Planes.Far => new Plane(x.w - x.z, y.w - y.z, z.z - z.y, w.w - w.z),
+                Planes.Left => new Plane(x.w + x.x, y.w + y.x, z.w + z.x, w.w + w.x),
+                Planes.Top => new Plane(x.w - x.y, y.w - y.y, z.w - z.y, w.w - w.y),
+                Planes.Right => new Plane(x.w - x.x, y.w - y.x, z.w - z.x, w.w - w.x),
+                Planes.Bottom => new Plane(x.w + x.y, y.w + y.y, z.w + z.y, w.w + w.y),
+            };
+            newPlane.Normal = -newPlane.Normal;
+            return newPlane.Normalized();
+        }
```

Comment:
added `_ => new Plane()`

1. normal is `Zero`
1. normal negated is `-zero` 
1. `normalized` checks if the normal is zero and returns `Plane` with all zero

The function does more work, but the result is the same. I think the cleaner look is better as the performance for the unusal case.

Reply:
I agree with the cleaner look, not sure if `_ => new Plane()` is necessary since structs get default initialized anyway. I originally missed that `normalized` would return a `Plane` with all zero, I was worried we'd be returning negative zeroes.

Reply:
This type of switch throws a `SwitchExpressionException` if a case is not handled. The `_ => new Plane()` is required to match the c implementation.